### PR TITLE
Integrated rating and feedback

### LINKED ISSUE #2181: Integrate review and feedback
**Expected Behaviour**

screen shot 2018-06-16 at 11 44 58 pm

On clicking rate this product,

screen shot 2018-06-17 at 12 00 09 am

This would require API to support review rating + feedback support

Cases:

1. Rating is must, feedback is optional. If a user clicks sends after sending a rating it should send only rating.

2. Title is optional

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by Dilshaad21**
Fixes #2181 

Changes: 
- Integrated rating and feedback.
- Rating is necessary but the feedback is optional 

To be merged after https://github.com/fossasia/susi_server/pull/1451 
 
Demo Link: https://pr-3442-fossasia-susi-web-chat.surge.sh

Screenshots of the change: 

final_5e4003f4823bf400151766b0_645856

Screenshot from 2020-02-09 18-14-13

Screenshot from 2020-02-09 18-51-49

### CODE REVIEW

**Comment by Dilshaad21** [comment]
WIP just need to make minute changes once PR in server side is merged

File: src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js [hunk 1]
```diff
@@ -224,6 +254,18 @@ class SkillFeedbackCard extends Component {
                     : `${userEmail.slice(0, userEmail.indexOf('@') + 1)}...`}
                 </div>
                 <Timestamp>{formatDate(parseDate(data.timestamp))}</Timestamp>
+                <Ratings
+                  rating={3}
```
Comment (Dilshaad21) [inline src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: Just need to replace `rating` with `data.stars`. It's hardcoded to 3 for now

Comment (snitin315) [reply src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: Do this change here are edit the PR description with , 
TO BE MERGED AFTER #SERVER_PR

Comment (Dilshaad21) [reply src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: If I do that change now the checks will fail and it will not get deployed to surge.

**Comment by Dilshaad21** [comment]
@akshatnitd PTAL

**Comment by Dilshaad21** [comment]
@akshatnitd PTAL

File: src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js [hunk 2]
```diff
@@ -242,45 +345,72 @@ class SkillFeedbackCard extends Component {
...
-                <Button
-                  label="Post"
-                  color="primary"
-                  variant="contained"
-                  style={{ marginTop: 10 }}
-                  handleClick={this.postFeedback}
-                  disabled={newFeedbackValue.trim().length === 0 || loading}
-                  isLoading={loading}
-                  buttonText="Post"
+            <div
```
Comment (anishagg17) [inline src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: It can be transformed to styled component

Comment (Dilshaad21) [reply src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: I think it's okay to use inline styling for one-two styling attributes.

**Comment by Dilshaad21** [comment]
@akshatnitd changes have been made
- Implemented it to skillFeedbackPage.
- Handled the delay in displaying self feedback

File: src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js [hunk 3]
```diff
@@ -81,31 +86,89 @@ class SkillFeedbackCard extends Component {
...
   };
 
-  postFeedback = async () => {
+  editRating = rating => {
+    this.setState({ rating });
+  };
+
+  getRatingMessage = () => {
+    const { rating } = this.state;
+    switch (rating) {
```
[the remark this answers is not in the document]
Comment (akshatnitd) [reply src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: `RATING_TO_HINT_MAP`

File: src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js [hunk 4]
```diff
@@ -233,6 +259,20 @@ class SkillFeedbackCard extends Component {
                       : `${userEmail.slice(0, userEmail.indexOf('@') + 1)}...`}
                   </div>
                   <Timestamp>{formatDate(parseDate(data.timestamp))}</Timestamp>
+                  {data.rating > 0 ? (
+                    <Ratings
+                      rating={data.rating}
```
Comment (akshatnitd) [inline src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: Move it to a common place, it is repeated at multiple places

Comment (Dilshaad21) [reply src/components/cms/SkillFeedbackCard/SkillFeedbackCard.js]: Made a separate `Rating` component for that.

File: src/components/shared/Rating.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import Ratings from 'react-ratings-declarative';
+
+const Rating = props => {
```
Comment (akshatnitd) [inline src/components/shared/Rating.js]: Use better name. Eg - `FiveStarRatingWidget`

File: src/components/cms/SkillFeedbackPage/SkillFeedbackPage.js [hunk 6]
```diff
@@ -540,18 +615,37 @@ class SkillFeedbackPage extends Component {
...
                 <FormControl fullWidth={true}>
+                  <div
+                    style={{
+                      textAlign: 'center',
+                      margin: '1.5rem',
+                    }}
+                  >
+                    <Rating
+                      rating={this.state.rating}
+                      widgetRatedColors="#ffbb28"
```
Comment (akshatnitd) [inline src/components/cms/SkillFeedbackPage/SkillFeedbackPage.js]: These props are present in almost all scenarios. Please put them under the common component.

File: src/components/cms/SkillFeedbackPage/EditFeedbackDialog.js [hunk 7]
```diff
@@ -6,15 +6,35 @@ import DialogTitle from '@material-ui/core/DialogTitle';
...
+    handleEditFeedback,
+    handleEditRating,
+    rating,
+  } = props;
 
   const [loader, updateLoader] = useState(false);
   const [newFeedback, updateNewFeedback] = useState(feedback);
+  const [newRating, updateRating] = useState(rating);
+  const mobileView = isMobileView();
+  const RATING_TO_HINT_MAP = new Map([
```
Comment (akshatnitd) [inline src/components/cms/SkillFeedbackPage/EditFeedbackDialog.js]: Put it under constant folder. And reuse it.

Use syntax similar to usage across repo - 
```
const RATING_TO_HINT_MAP = {
0: 'The skill has not been rated by you',
1: 'Hated it',
...
};
```

File: src/components/shared/FiveStarRatingWidget.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  widgetDimensions: PropTypes.string,
+  widgetSpacings: PropTypes.string,
+  rating: PropTypes.number,
+};
+
+FiveStarRatingWidget.defaultProps = {
+  widgetRatedColors: '#ffbb28',
+  widgetDimensions: '20px',
+  widgetSpacings: '0px',
+};
```
Comment (Dilshaad21) [inline src/components/shared/FiveStarRatingWidget.js]: @akshatnitd I have initialised these common props using `defaultProps`, now the rating widget in the feedback card need to pass only `rating` as props except the widget which involves `changeRating` which which uses some other styles such as `widgetHoverColor` that can be passed as well and will be accessed from `otherProps`.

**Pull request merged**